proposed/2021/DotNetAudit.md

```diff
@@ -0,0 +1,296 @@
+          ]
+        }
+      ]
+    },
+    "outdated": {
+      "packages": "2",
+      "dependencies": [
+        {
+          "name": "anurse.testing.TestDeprecatedPackage",
+          "resolvedVersion": "1.0.0",
+          "latestVersion": "2.0.0"
+        },
+        {
+          "name": "UmbracoForms",
+          "resolvedVersion": "8.4.1",
+          "latestVersion": "8.7.1"
+        }
+      ]
+    },
+    "scannedDependencies": "36"
+  }
+}
+```
+
+##### dotnet audit Exit Codes
+
+- 0 - The command will exit with a 0 exit code if no vulnerabilities, deprecations, or outdated packages were found.
+- 1 - The command will exit with a 1 exit code if a vulnerability, deprecation, or outdated package was found.
+- 2 - The command will exit with a 2 exit code if it unexpectedly failed.
+- 3 - The command will exit with a 3 exit code if an unsupported project is detected.
+
+#### dotnet audit fix
+
+The `dotnet audit fix` command will provide a remediation that is calculated with an implicit `dotnet audit` to then apply directly to a resulting package graph. It can add packages, remove packages, and update packages depending on the problem it's attempting to resolve. It does not take into consideration downgrading to a compatible version if a higher one has already been specified.
+
+![dotnet audit fix](/meta/resources/dotnetaudit/dotnetauditfix.png)
+
+```
+$ dotnet audit fix
+
+Fixing vulnernable packages in ContosoUniversity.sln
+	Upgrading UmbracoForms 8.4.1 to 8.7.1
+
+Fixing deprecated packages in ContosoUniversity.sln
+	Replacing anurse.testing.TestDeprecatedPackage 1.0.0 with Microsoft.AspNetCore.Mvc 2.2.0
+
+Fixing outdated packages in ContosoUniversity.sln
+	Packages are currently up-to-date.
+    
+Fixed 2 packages in 36 scanned packages.
+```
+
+##### dotnet audit fix --dry-run
+
+Does a dry run to give an idea of what `audit fix` will do. Provides output, but does not commit the fix.
+
+##### dotnet audit fix --json
+
+```json
+  "added": [
+    {
+      "name": "Microsoft.AspNetCore.MVC",
+      "version": "2.2.0"
+    }
+  ],
+  "removed": [
+      {
+        "name": "anurse.testing.TestDeprecatedPackage",
+        "version": "1.0.0"
+      }
+  ],
+  "updated": [
+    {
+      "name": "UmbracoForms",
+      "version": "8.7.1",
+      "previousVersion": "8.4.1"
+    }
+  ],
+  "failures": [],
+  "warnings": []
+```
+
+##### dotnet audit fix Exit Codes
+
+- 0 - The command will exit with a 0 exit code if no vulnerabilities, deprecations, or outdated packages were found *or* remediation was able to fix all issues.
+- 1 - The command will exit with a 1 exit code if a vulnerability, deprecation, or outdated package was found *and* remediation is not able to fix all issues.
+- 2 - The command will exit with a 2 exit code if it unexpectedly failed.
+- 3 - The command will exit with a 3 exit code if an unsupported project is detected.
+
+#### Vulnerabilities
+
+When vulnerable packages are detected, an error is thrown by default.
```

Comment:
Could you explain why an error is thrown instead of a warning for vulnerabilities?

A reason I ask is because vulnerabilities in the Javascript ecosystem often noise or very low impact.

Reply:
I'll add more rationale in the proposal. I believe with our lower amount of known vulnerabilities, high importance of the types of vulnerabilities issued for .NET, and the nature of shifting left, these should be treated as errors for now.

https://overreacted.io/npm-audit-broken-by-design/